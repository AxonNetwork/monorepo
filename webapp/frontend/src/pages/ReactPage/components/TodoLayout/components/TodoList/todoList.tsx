import { List } from 'immutable';
import React from 'react';
import { connect } from 'react-redux';
import { AnyAction, Dispatch } from 'redux';

import { toggleTodo } from 'services/todos/actions';
import { VISIBILITY_FILTER_OPTIONS } from 'services/todos/constants';
import { ITodo, ITodosState } from 'services/todos/types';
import { IGlobalState } from 'types/global';
import Todo from './components/Todo';

// Component

interface ITodoListStateProps {
  todos: List<ITodo>;
}

interface ITodoListDispatchProps {
  toggleTodo(id: string): void;
}

interface ITodoListProps extends ITodoListStateProps, ITodoListDispatchProps { }

export class TodoList extends React.Component<ITodoListProps> {
  public onClick = (id: string) => {
    return () => {
      this.props.toggleTodo(id);
    };
  }

  public render() {
    const todoList = this.props.todos.map((todo) =>
      <Todo key={todo.id} {...todo} onClick={this.onClick(todo.id)} />,
    );
    if (todoList.count() > 0) {
      return (
        <ul className='collection'>
          {todoList}
        </ul>
      );
    } else {
      return (null);
    }
  }
}

// Container

const getVisibleTodos = (todosState: ITodosState): List<ITodo> => {
  switch (todosState.visibilityFilter) {
    case VISIBILITY_FILTER_OPTIONS.SHOW_ALL:
      return todosState.todos;
    case VISIBILITY_FILTER_OPTIONS.SHOW_COMPLETED:
      return todosState.todos.filter((t) => t !== undefined && t.completed).toList();
    case VISIBILITY_FILTER_OPTIONS.SHOW_ACTIVE:
      return todosState.todos.filter((t) => t !== undefined && !t.completed).toList();
    default:
      return todosState.todos;
  }
};

const mapStateToProps = (state: IGlobalState): ITodoListStateProps => ({
  todos: getVisibleTodos(state.todosState),
});

const mapDispatchToProps = (dispatch: Dispatch<AnyAction>): ITodoListDispatchProps => ({
  toggleTodo: (id: string) => {
    dispatch(toggleTodo(id));
  },
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(TodoList);
