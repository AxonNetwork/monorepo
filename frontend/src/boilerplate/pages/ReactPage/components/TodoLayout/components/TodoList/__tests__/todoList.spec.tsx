import { List } from 'immutable';
import * as React from 'react';
import * as TestRenderer from 'react-test-renderer';

import { TodoList } from '../todoList';

jest.mock('../components/Todo', () => ({ default: 'Todo' }));
const dispatchMock = () => { return; };

describe('TodoList', () => {
  it('should renders correctly', () => {
    const renderer = TestRenderer.create(
      <TodoList todos={List<any>()} toggleTodo={dispatchMock} />,
    );
    expect(renderer).toMatchSnapshot();
    renderer.unmount();
  });
});
