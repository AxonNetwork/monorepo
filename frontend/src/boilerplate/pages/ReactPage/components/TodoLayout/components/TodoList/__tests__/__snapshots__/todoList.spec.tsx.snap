// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList should renders correctly 1`] = `null`;
